nixos/modules/services/monitoring/prometheus/exporters/fritz.nix

```diff
@@ -0,0 +1,98 @@
+{ config, lib, pkgs, options }:
+let
+  inherit (lib) mkOption types mdDoc concatStringsSep;
+  cfg = config.services.prometheus.exporters.fritz;
+  yaml = pkgs.formats.yaml { };
+  configFile = yaml.generate "fritz-exporter.yaml" cfg.settings;
+in
+{
+  port = 9787;
+
+  extraOpts = {
+    settings = mkOption {
+      description = mdDoc "Configuration settings for fritz-exporter.";
+      type = types.submodule {
+        freeformType = yaml.type;
+
+        options = {
+          # Pull existing port option into config file.
+          port = mkOption {
+            type = types.port;
+            default = cfg.port;
+            internal = true;
+            visible = false;
+          };
+          # Pull existing listen address option into config file.
+          listen_address = mkOption {
+            type = types.str;
+            default = cfg.listenAddress;
+            internal = true;
+            visible = false;
+          };
+          log_level = mkOption {
+            type = types.enum [ "DEBUG" "INFO" "WARNING" "ERROR" "CRITICAL" ];
+            default = "INFO";
+            description = mdDoc ''
+              Log level to use for the exporter.
+            '';
+          };
+          devices = mkOption {
+            default = [];
+            description = "Fritz!-devices to monitor using the exporter.";
+            type = with types; listOf (submodule {
+              freeformType = yaml.type;
+
+              options = {
+                name = mkOption {
```

Comment:
with this module setup and the yaml writer the name is currently not optional.
Not setting the value writes `name: null` to the config file,

```nix
services = {
    prometheus.exporters.fritz = {
      enable = true;
      listenAddress = "127.0.0.1";
      settings.devices = [{
        username = "prometheus";
        password_file = secrets.prometheus-fritzbox.path;
      }];
    };
};
```

```
╰─ cat /nix/store/b28k8bzsc1s1lq7jsydvx82yq4kvjdzi-fritz-exporter.yaml
devices:
- host_info: false
  hostname: fritz.box
  name: null
  password_file: /run/secrets/prometheus-fritzbox
  username: prometheus
listen_address: 127.0.0.1
log_level: INFO
port: 9787


```

 with results the exporter to crash with

```
Mar 17 22:58:51 tank fritzexporter[1510179]: 2024-03-17 22:58:51,923     INFO fritzexporter | Using password from password file /run/secrets/prometheus-fritzbox
Mar 17 22:58:56 tank fritzexporter[1510179]: 2024-03-17 22:58:56,494     INFO fritzexporter.fritzdevice | Connection to fritz.box successful, reading capabilities
Mar 17 22:58:59 tank fritzexporter[1510179]: 2024-03-17 22:58:59,165     INFO fritzexporter.fritzdevice | Reading capabilities for fritz.box, got serial 444E6D43E718, model name FRITZ!Box 7590 completed
Mar 17 22:58:59 tank fritzexporter[1510179]: 2024-03-17 22:58:59,165     INFO fritzexporter | registering fritz.box to collector
Mar 17 22:59:03 tank fritzexporter[1510179]: 2024-03-17 22:59:03,146     INFO fritzexporter | Starting listener at 127.0.0.1:9787
Mar 17 22:59:03 tank fritzexporter[1510179]: 2024-03-17 22:59:03,147     INFO fritzexporter | Entering async main loop - exporter is ready
Mar 17 22:59:07 tank fritzexporter[1510179]: Traceback (most recent call last):
Mar 17 22:59:07 tank fritzexporter[1510179]:   File "/nix/store/3v2ch16fkl50i85n05h5ckss8pxx6836-python3-3.11.8/lib/python3.11/wsgiref/handlers.py", line 137, in run
Mar 17 22:59:07 tank fritzexporter[1510179]:     self.result = application(self.environ, self.start_response)
Mar 17 22:59:07 tank fritzexporter[1510179]:                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 17 22:59:07 tank fritzexporter[1510179]:   File "/nix/store/20ygb85264lphxzfrcaqv00gkd4ghigd-python3.11-prometheus-client-0.19.0/lib/python3.11/site-packages/prometheus_client/exposition.py", line 128, in prometheus_app
Mar 17 22:59:07 tank fritzexporter[1510179]:     status, headers, output = _bake_output(registry, accept_header, accept_encoding_header, params, disable_compression)
Mar 17 22:59:07 tank fritzexporter[1510179]:                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 17 22:59:07 tank fritzexporter[1510179]:   File "/nix/store/20ygb85264lphxzfrcaqv00gkd4ghigd-python3.11-prometheus-client-0.19.0/lib/python3.11/site-packages/prometheus_client/exposition.py", line 104, in _bake_output
Mar 17 22:59:07 tank fritzexporter[1510179]:     output = encoder(registry)
Mar 17 22:59:07 tank fritzexporter[1510179]:              ^^^^^^^^^^^^^^^^^
Mar 17 22:59:07 tank fritzexporter[1510179]:   File "/nix/store/20ygb85264lphxzfrcaqv00gkd4ghigd-python3.11-prometheus-client-0.19.0/lib/python3.11/site-packages/prometheus_client/exposition.py", line 282, in generate_latest
Mar 17 22:59:07 tank fritzexporter[1510179]:     output.append(sample_line(s))
Mar 17 22:59:07 tank fritzexporter[1510179]:                   ^^^^^^^^^^^^^^
Mar 17 22:59:07 tank fritzexporter[1510179]:   File "/nix/store/20ygb85264lphxzfrcaqv00gkd4ghigd-python3.11-prometheus-client-0.19.0/lib/python3.11/site-packages/prometheus_client/exposition.py", line 239, in sample_line
Mar 17 22:59:07 tank fritzexporter[1510179]:     ['{}="{}"'.format(
Mar 17 22:59:07 tank fritzexporter[1510179]:   File "/nix/store/20ygb85264lphxzfrcaqv00gkd4ghigd-python3.11-prometheus-client-0.19.0/lib/python3.11/site-packages/prometheus_client/exposition.py", line 240, in <listcomp>
Mar 17 22:59:07 tank fritzexporter[1510179]:     k, v.replace('\\', r'\\').replace('\n', r'\n').replace('"', r'\"'))
Mar 17 22:59:07 tank fritzexporter[1510179]:        ^^^^^^^^^

```

When setting a name with in the module, the error goes away.
So we  either should make it not write the `null` value (not sure if that is possible )or enforce the user to set it to something.

Beside that small thing, the PR works fine on my end. :slightly_frowning_face:

Reply:
Oh, thats a problem, I'm not sure what would be the best way to fix this

Reply:
I think defaulting to an empty string instead would be our best option, as that is also upstreams default: https://github.com/pdreker/fritz_exporter/blob/413586d8765cfd3a300c8dd6c73e0d281723e860/fritzexporter/config/config.py#L168